### PR TITLE
feat: enable ssh tunneling using env variables for superset

### PR DESCRIPTION
**Pull request by njuguna-n**
#56 

Enable SSH tunneling for superset

### CODE REVIEW

**Review by lorerod — CHANGES_REQUESTED** [review]
Hi @njuguna-n, thank you for these changes.
Please add steps on the README file to explain setting up SSH tunneling.
I also left some comments inline.

**Comment by mrjones-plip** [comment]
Hey all!  Amazing timing on this PR!  I was just working on [another superset deployment](https://github.com/medic/medic-infrastructure/issues/713), and didn't know that there was built in support for SSH tunnels.  Thank you for the hard work and PR here <3

What about more standard docker compose  based deployment of Superset? There might be some history I'm missing which could make me appreciate why we're chose to use a bespoke compose solution for CHT Sync's Superset.  I'm for sure coming into this late in the game, so there may be some real good reasons we're here!

For example, I notice our [compose file](https://github.com/medic/cht-sync/blob/56-enable-ssh-tunneling/docker-compose.yml) is limited as compared to the [Superset's production one](https://superset.apache.org/docs/installation/installing-superset-using-docker-compose/#3-launch-superset-through-docker-compose) (aka `docker-compose-non-dev.yml`). I believe this is forcing us to compile a local image which [statically includes](https://github.com/medic/cht-sync/blob/febfd4ff596349d858ed2a30e511e47f0348a2ed/superset/Dockerfile#L9) the `superset_config.py` file.  If we use Superset's we get a [bind mounted version](https://github.com/apache/superset/tree/master/docker#local-overrides) which might avoid compiling. The `SECRET_KEY` and the is set in [their example production `.env` file](https://github.com/apache/superset/blob/master/docker/.env-non-dev), which we're not [leveraging either](https://github.com/medic/cht-sync/blob/56-enable-ssh-tunneling/env.template).  

I'm concerned about the lack of a caching layer, like Redis.  When I worked on the Covid MoH Mali Superset dashboard, there was a point when we realized the system was so slow because there was no cache - likely it was falling back to filesystem cache or, worse, no cache (I failed to find Superset's default :shrug: ).

**Comment by mrjones-plip** [comment]
Oh - I do see that production install flow is "quick, login with `admin` / `admin` and change you're password" which is kind of a loser move for an internet exposed instance.  Fix it in our docs?

File: superset/superset_config.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import os
+
+FEATURE_FLAGS = {
+    "ENABLE_TEMPLATE_PROCESSING": True,
+    "SSH_TUNNELING": os.environ["SUPERSET_SSH_TUNNELING"],
+}
+
+ENABLE_PROXY_FIX = True
+SECRET_KEY = "YOUR_OWN_RANDOM_GENERATED_STRING"
```
[the remark this answers is not in the document]
Comment (njuguna-n) [reply superset/superset_config.py:9]: Yes, that should also be set with an env variable. I will make the change.

File: env.template [hunk 2]
```diff
@@ -26,3 +26,4 @@ COUCHDB_SECURE=false
 # superset: required environment variables for 'gamma', 'prod' and 'local'
 SUPERSET_PASSWORD=password
 SUPERSET_ADMIN_EMAIL=user@superset.com
+SUPERSET_SSH_TUNNELING=false
```
[the remark this answers is not in the document]
Comment (njuguna-n) [reply env.template:29]: Thanks for catching that. I will make the update.

**Comment by njuguna-n** [comment]
@mrjones-plip I was created a [ticket](https://github.com/medic/cht-pipeline/issues/58) today specifically to handle how we set up superset especially for production so your comment has come at a handy time as well 😄. The initial idea of adding the superset image the way we did was to limit the number of Docker containers required to get cht-sync up and running, we already have 8 containers that are used in the dev environment. This approach definitely needs to change for production as you rightly pointed out

**Comment by mrjones-plip** [comment]
oh - that totally makes sense @njuguna-n  to have a simplified dev environment - thanks!  I suspect we can follow Superset's lead and have dev and prod compose files an environment files.  Lemme know if you need a hand.

Also - I just tested with vanilla `2.1.2` docker compose production install it looks like maybe SSH Tunnels are supported with no extra config? I think maybe due to a [change in the default](https://github.com/apache/superset/pull/22723#issue-1532615753). Here's my test in the GUI ( but didn't actual test a live tunnel):


https://github.com/medic/cht-sync/assets/8253488/70c1a6fc-8f8b-48b2-a96a-f8f2801c1e3d

**Comment by njuguna-n** [comment]
Is the SHH Tunnel option still there if you click on the `SQLAlchemy URI string` option?

**Comment by njuguna-n** [comment]
> I suspect we can follow Superset's lead and have dev and prod compose files an environment files. 
Yes, it makes sense to follow do that. Do you think we still need a cache for superset on the dev instance?

**Comment by njuguna-n** [comment]
There should be a toggle right above the `Test Connection` button. Maybe the default does not work. I am using this branch which has version `2.1.2` as well.

https://github.com/medic/cht-sync/assets/141340177/a159d9a0-dbe7-4d0d-b627-f7fad759ea6a

**Comment by mrjones-plip** [comment]
@njuguna-n - we got the tunnel working, but for sure had to use the ` superset_config_docker.py` trick (even in Superset 3.x).  [Here's the steps](https://github.com/medic/medic-infrastructure/issues/713#issuecomment-1843265475) we took, and if you don't do that, you see this:

image

Still no testing of the `SQLAlchemy URI string` option - but happy to jump on a call and test as I have SSH and Postgres credentials that work!

**Comment by njuguna-n** [comment]
@mrjones-plip I pushed a commit to change how the superset container is created. Please have a look.

**Comment by mrjones-plip** [comment]
@njuguna-n - wow!  Some really nice work here at improving CHT Sync to not build custom images, use version pinned upstream images of all the important services and use docker compose inheritance to chain together multiple configs.  Excellent!

I trust the teams testing on this, so I haven't dived into running the code, but I did notice one possible error here in the readme in prod where you have an [optional step 2](https://github.com/medic/cht-sync/blob/9b923b35e6b902f1a468f957afa9b582a5b9a704/README.md?plain=1#L163) of :

```
docker-compose -f docker-compose.postgres.yml -f docker-compose.yml up postgres
```

Two changes I think are needed:
1. add a trailing ` -d` to make it run in the background like the rest of the `up` calls
2. remove `f docker-compose.yml` as this [will start](https://github.com/medic/cht-sync/blob/56-enable-ssh-tunneling/docker-compose.yml) logstash and dbt which we start in the `npm` [calls](https://github.com/medic/cht-sync/blob/9ad0c419847d420814a85e0504e82141716ee030/scripts/config.ts#L44).

**Review by lorerod — CHANGES_REQUESTED** [review]
Can you please update the e2e test to successfully connect to the superset instance?

**Comment by njuguna-n** [comment]
Closing this PR since we have decided not to include Superset in this repo for now. More context on this in [slack](https://medic.slack.com/archives/C03SP4KRVEX/p1705079537536179).

**Pull request closed without merging**